### PR TITLE
Do not update allowedWorklodsPerQueue directly in AddClusterQueue

### PR DESCRIPTION
**Pull request by tenzen-y**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
The `admittedWorkloadsPerQueue` uses the queue's name as a key only in `func AddClusterQueue`, even though it uses queueKey (Namespace/Name) as a key in other places for the cache.

Also, I changed the type of `admittedWorkloadsPerQueue` from `int` to `QUEUE_KEY: sets.Set[WORKLOAD_KEY]` to avoid being double counted workloads in `AddLocalQueue` and `AddClusterQueue`.

https://github.com/kubernetes-sigs/kueue/blob/9ca57c86cf06c11a94a2d5b7badf60233a51a2f2/pkg/cache/cache.go#L549-L560

https://github.com/kubernetes-sigs/kueue/blob/9ca57c86cf06c11a94a2d5b7badf60233a51a2f2/pkg/cache/cache.go#L420-L426

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

**Comment by tenzen-y** [comment]
Probably, we should add more tests to https://github.com/kubernetes-sigs/kueue/blob/9ca57c86cf06c11a94a2d5b7badf60233a51a2f2/pkg/cache/cache_test.go...

File: pkg/cache/cache.go [hunk 1]
```diff
@@ -552,8 +552,9 @@ func (c *Cache) AddClusterQueue(ctx context.Context, cq *kueue.ClusterQueue) err
 			continue
 		}
 		c.addOrUpdateWorkload(&workloads.Items[i])
-		if _, ok := cqImpl.admittedWorkloadsPerQueue[w.Spec.QueueName]; ok {
-			cqImpl.admittedWorkloadsPerQueue[w.Spec.QueueName]++
+		qKey := workload.QueueKey(&w)
+		if _, ok := cqImpl.admittedWorkloadsPerQueue[qKey]; ok {
```
Comment (alculquicondor) [inline pkg/cache/cache.go:562]: oops, can you add a unit test for this?

**Review by alculquicondor — COMMENTED** [review]
Did you notice a case of double counting or the change to set is preventative?

**Comment by tenzen-y** [comment]
> Did you notice a case of double counting or the change to set is preventative?

@alculquicondor I faced the double counting case at https://github.com/kubernetes-sigs/kueue/blob/9ca57c86cf06c11a94a2d5b7badf60233a51a2f2/pkg/cache/cache_test.go#L1432/

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/714/pull-kueue-test-unit-main/1650255668571017216

File: pkg/cache/cache.go [hunk 2]
```diff
@@ -552,8 +558,9 @@ func (c *Cache) AddClusterQueue(ctx context.Context, cq *kueue.ClusterQueue) err
 			continue
 		}
 		c.addOrUpdateWorkload(&workloads.Items[i])
```
Comment (alculquicondor) [inline pkg/cache/cache.go:560]: doesn't this already handle the localQueue counters? Can we just remove the lines below?

Comment (tenzen-y) [reply pkg/cache/cache.go:560]: IIUC, this function doesn't update `admittedWorkloadsPerQueue`, right?
Do you mean, we don't need to update `admittedWorkloadsPerQueue `?

Comment (alculquicondor) [reply pkg/cache/cache.go:560]: I think it does, indirectly.

Comment (alculquicondor) [reply pkg/cache/cache.go:560]: That might be why the unit tests were passing before this PR

Comment (tenzen-y) [reply pkg/cache/cache.go:560]: Ah. I see.

**Comment by tenzen-y** [comment]
/retitle Do not update allowedWorklodsPerQueue directly in AddClusterQueue

File: pkg/cache/cache.go (hunk 1 above)
Comment (tenzen-y) [reply pkg/cache/cache.go:562]: Probably, we don't need to add a unit test anymore.

File: pkg/cache/cache.go (hunk 2 above)
Comment (tenzen-y) [reply pkg/cache/cache.go:560]: done.

Comment (alculquicondor) [reply pkg/cache/cache.go:560]: then... do we still need the set or can we just live with the integer?

Comment (tenzen-y) [reply pkg/cache/cache.go:560]: Uhm... If we use the set, we can avoid the potential issues. Although, we may be able to keep using integer for now.
We can come back to this discussion after we face this issue again.

Comment (tenzen-y) [reply pkg/cache/cache.go:560]: This means I will revert it.

Comment (tenzen-y) [reply pkg/cache/cache.go:554]: done.

Comment (alculquicondor) [reply pkg/cache/cache.go:554]: Thanks. I prefer we avoid an extra set for performance reasons.

**Comment by tenzen-y** [comment]
I wonder why this PR isn't merged into the main by prow.

**Comment by tenzen-y** [comment]
> tide — Not mergeable. Merge type "merge" disallowed by repo settings

@alculquicondor Did you change the repo settings?

**Comment by alculquicondor** [comment]
oh, I changed the wrong setting :(

**Comment by alculquicondor** [comment]
can you force push to see if it gets fixed?

**Comment by tenzen-y** [comment]
> can you force push to see if it gets fixed?

@alculquicondor Forced push didn't fix this situation :(

**Comment by alculquicondor** [comment]
Let's wait for the tests to finish

**Comment by tenzen-y** [comment]
> Let's wait for the tests to finish

Ah, I see.

**Comment by alculquicondor** [comment]
I give up... could you open a new PR?

**Comment by tenzen-y** [comment]
> I give up... could you open a new PR?

Sure.
/close

**Pull request closed without merging**